packages/@aws-cdk/aws-iam/test/merge-statements.test.ts

```diff
@@ -0,0 +1,328 @@
+    }),
+    new iam.PolicyStatement({
+      resources: ['b'],
+      actions: ['service:Action'],
+      principals: [principal],
+    }),
+    // This can combine with the previous two once they have been merged
+    new iam.PolicyStatement({
+      resources: ['a', 'b'],
+      actions: ['service:Action2'],
+      principals: [principal],
+    }),
+  ], [
+    {
+      Effect: 'Allow',
+      Resource: ['a', 'b'],
+      Action: ['service:Action', 'service:Action2'],
+      Principal: { AWS: PRINCIPAL_ARN },
+    },
+  ]);
+});
+
+test('merges 2 pairs separately', () => {
+  // Merges pairs (0,2) and (1,3)
+  assertMerged([
+    new iam.PolicyStatement({
+      resources: ['a'],
+      actions: ['service:Action'],
+      principals: [principal],
+    }),
+    new iam.PolicyStatement({
+      resources: ['c'],
+      actions: ['service:Action1'],
+      principals: [principal],
+    }),
+    new iam.PolicyStatement({
+      resources: ['b'],
+      actions: ['service:Action'],
+      principals: [principal],
+    }),
+    new iam.PolicyStatement({
+      resources: ['c'],
+      actions: ['service:Action2'],
+      principals: [principal],
+    }),
+  ], [
+    {
+      Effect: 'Allow',
+      Resource: ['a', 'b'],
+      Action: 'service:Action',
+      Principal: { AWS: PRINCIPAL_ARN },
+    },
+    {
+      Effect: 'Allow',
+      Resource: 'c',
+      Action: ['service:Action1', 'service:Action2'],
+      Principal: { AWS: PRINCIPAL_ARN },
+    },
+  ]);
+});
+
+test('do not deep-merge info Refs and GetAtts', () => {
+  const stack = new Stack();
+  const user1 = new iam.User(stack, 'User1');
+  const user2 = new iam.User(stack, 'User2');
+
+  assertMerged([
+    new iam.PolicyStatement({
+      resources: ['a'],
+      actions: ['service:Action'],
+      principals: [user1],
+    }),
+    new iam.PolicyStatement({
+      resources: ['a'],
+      actions: ['service:Action'],
+      principals: [user2],
+    }),
+  ], [
+    {
+      Effect: 'Allow',
+      Resource: 'a',
+      Action: 'service:Action',
+      Principal: {
+        AWS: [
+          { 'Fn::GetAtt': ['User1E278A736', 'Arn'] },
+          { 'Fn::GetAtt': ['User21F1486D1', 'Arn'] },
+        ],
+      },
+    },
+  ]);
+});
+
+function assertNoMerge(statements: iam.PolicyStatement[]) {
+  const app = new App();
+  const stack = new Stack(app, 'Stack');
+
+  const regularResult = stack.resolve(new iam.PolicyDocument({ minimize: false, statements }));
+  const minResult = stack.resolve(new iam.PolicyDocument({ minimize: true, statements }));
+
+  expect(minResult).toEqual(regularResult);
+}
+
+function assertMerged(statements: iam.PolicyStatement[], expected: any[]) {
+  const app = new App();
+  const stack = new Stack(app, 'Stack');
+
+  const minResult = stack.resolve(new iam.PolicyDocument({ minimize: true, statements }));
+
+  expect(minResult.Statement).toEqual(expected);
+}
+
+function assertMergedP(doMerge: boolean, statements: iam.PolicyStatement[], expected: any[]) {
```

Comment:
Sorry, what's `assertMergedP`? Not sure what the `P` stands for.

Maybe a better name is in order here?